Feature/61610 date picker to edit phases and phase gates of a project lifecycle (2/2)

dombesz opened the pull request:
# Ticket
https://community.openproject.org/wp/61610

# What are you trying to accomplish?
Change the project phases datepicker to be displayed in an editing modal, disable non working days and make it function more like the work package datepicker modal.

# What approach did you choose and why?
Todo:
- [X] Style the editing link to look like an input field when hovering it.
- [x] Split the date range input field into Start, Finish dates and Duration.
- [x] Display datepicker inside the the edit dialog instead of its own pop up.
- [x] Disable setting the Start date if a previous phase has a Finish date.
- [x] Implement highlighted field logic in order to allow choosing dates for the focused input field.
- [x] Disable selecting dates that are before the previous phase's finish date.
- [x] Add validation that no start and end date are on a Non-Working day.
- [x] Add/update specs.
- [x] Update mobile view to render system datepicker
- [x] Fix bug when sometimes the start date gets filled if no dates are selected and the finish date is focused.
        Have no start date, set finish date, clear the finish date with the button, choose a new finish date. The start date will be set instead of the finish date. Reason is, turbo preview kills the focused field because it re renders on the clearing.
- [x] When typing with a certain speed, characters get dismissed from the date fields due to turbo preview updates

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

dombesz committed:
Order sass imports alphabetically

dombesz committed:
[#61610] Split life cycle datepicker input fields into start, finish date and duration

https://community.openproject.org/work_packages/61610

dombesz committed:
Split the datepicker range field into start, finish date and duration.

dombesz committed:
Connect the datepicker to the input fields

dombesz committed:
Move errors from the date range to the start date.

dombesz committed:
Disable the start date when a previous phase defines a finish date.

dombesz committed:
Set the default start date on the phase if the previous phase has a finish date, wip on autofocusing on re-render.

toy commented on app/models/project/phase.rb:
```diff
@@ -84,7 +86,14 @@ def range_incomplete?
   def validate_date_range
-    errors.add(:date_range, :start_date_must_be_before_finish_date) if range_set? && (start_date > finish_date)
+    if range_set? && (start_date > finish_date)
+      date_field, error_message = if finish_date_changed?
+                                    %i[finish_date must_be_after_start_date]
+                                  else
+                                    %i[start_date must_be_before_finish_date]
+                                  end
+      errors.add(date_field, error_message)
```
I think it is simpler to do more directly:
```suggestion
      if finish_date_changed?
        errors.add(:finish_date, :must_be_after_start_date)
      else
        errors.add(:start_date, :must_be_before_finish_date)
      end
```

dombesz committed:
Make datepicker UI integration to work with highlighted fields and auto focusing missing date fields.

dombesz committed:
Validate start and finish dates do not fall on a non working day.

dombesz committed:
Fix unit specs

dombesz committed:
Update specs to handle multiple date input fields, restore preview form url instantly.

Restore the preview form url instantly so that if somebody presses the
save button while the preview is loading, the form submits to the update
action.

dombesz committed:
Always render the unparsed value in order to avoid a flickering input value.

dombesz committed:
Fix rubocop ABC error

dombesz committed:
Do not replace the active input field's value from the server response to avoid flickering of the input.

Since the preview action triggers multiple times when a user types
slowly, some characters could be lost between requests. In order to
avoid this, do not replace the active input fields value from the
server.

dombesz committed:
Increase gap between the inputs to match the work package datepicker, hide the datepicker on mobile.

dombesz committed:
Display native datpickers on mobile devices, except safari.

dombesz committed:
Maintain the focused class on the input field between morphing responses.

dombesz committed:
Clear start date selection when the chosen finish date is earlier than the start date.

dombesz committed:
Fix typo

dombesz committed:
Add specs for project phase datepicker interaction

dombesz committed:
Avoid double rendering the modal overlay on the project phase form.

dombesz committed:
Fix bug validation message is not shown when a following phase has a default start date and an invalid finsih date

ulferts commented on lookbook/previews/open_project/projects/phase_component_preview.rb:
```diff
@@ -35,7 +35,7 @@ def phase_with_both_gates
                                  definition: definition(start_gate: true, finish_gate: true),
                                  start_date: Date.current - 2.days,
-                                 end_date: Date.current + 2.days)
+                                 finish_date: Date.current + 2.days)
```
Thanks for fixing those here

ulferts commented on app/services/project_life_cycle_steps/set_attributes_service.rb:
```diff
@@ -29,6 +29,7 @@
 module ProjectLifeCycleSteps
   class SetAttributesService < ::BaseServices::SetAttributes
     def perform(*)
+      model.set_default_start_date
```
This would normally be placed in

```
def set_default_attributes(_params)
  self.start_date = model.default_start_date
end
```

which would allow to remove the method from the model.

Ideally, the `default_start_date` method would also be moved to the contract since it is the contract's objective to enforce, and by that also to know about, constraints. But that might be harder to do since it is also referred to by the form.

dombesz replied:
Unfortunately using `set_default_attributes` attributes is not ideal, because it is executed for new objects only.
I created a new callback [`ensure_default_attributes`](https://github.com/opf/openproject/blob/64712f0ee03341161b74396458801e2bf8049455/app/services/base_services/set_attributes.rb#L59-L68) that is executed for both existing and new objects.

ulferts commented on app/models/project/phase.rb:
```diff
@@ -88,7 +90,13 @@ def range_incomplete?
 
   def validate_date_range
-    errors.add(:date_range, :start_date_must_be_before_finish_date) if range_set? && (start_date > finish_date)
+    if range_set? && (start_date > finish_date)
```
Not directly related to the code here but rather to the other methods like `#date_range=` and `#range_set?`. I am wondering if we still need them at all or if they should at least be reworked. The UI is now no longer range based (except for showing). The form sends start and finish date except so it feels weird having the range based methods in here.

dombesz replied:
Good point, I removed the `date_range=` as it is not being used anymore, however the `range_set?` is in use, I renamed it to `date_range_set?` as it seems like a better method name.

ulferts commented on frontend/src/stimulus/controllers/form-preview.controller.ts:
```diff
@@ -41,8 +41,10 @@ export default class FormPreviewController extends Controller<HTMLFormElement> {
 
     const form = this.element;
+    const submitUrl = form.action;
     form.action = this.urlValue;
```
I guess this line can now be deleted as the value is set anew a few lines below.

dombesz replied:
This supposed to restore the original submit url after the preview is submitted. Ideally this is not necessary because the turbo response replaces the url automatically. However, it is required for solving an edge case when the preview response is loaded slowly and meanwhile submit button still can be pressed.

Not replacing the url would lead to submitting the form to the preview url again instead of saving it.

ulferts commented on spec/contracts/project_life_cycle_steps/update_contract_spec.rb:
```diff
@@ -137,6 +137,71 @@ def create_phase(**) = create(:project_phase, project:, **)
       end
     end
+
+    describe "#validate_dates_must_be_on_working_days" do
```
Since the method is not actually called, I find that this breaks the black box character of the tests. I would therefore change this to either a flat listing without a describe  block to wrap it in or to use a context block in order to group the elements semantically, but not by method name. So, e.g. 

```
context "(non) working day start and finish dates" do
  ...tests...
end
```

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+  }
+
+  private toDate(date:string|null):Date|null {
+    if (date && /^\d{4}-\d{2}-\d{2}$/.test(date)) {
```
You could use the timezoneService here via:
```
const date = this.timezoneService.parseISODate(date)
if (date.isValid()) {
  return date.toDate();
} else {
  return null;
}
```

Admittedly, it does not safe any lines. The only benefit I see is that you don't repeat the date format here. Not sure if that is worth it.

dombesz replied:
The problem is, moment.js is too permissive and it would parse even the following dates, and that is not desired in this case. (The work package form datepicker uses the same logic implemented here)

```js
moment("2021-01-213", "YYYY-MM-DD").isValid()
true
moment.utc("2021-01-213", "YYYY-MM-DD").isValid()
true
moment.utc("2021-01-213a", "YYYY-MM-DD").isValid()
true
moment.utc("2021-01a-213a", "YYYY-MM-DD").isValid()
true
moment.utc("2021-01a-213a", "YYYY-MM-DD").toDate()
Date Thu Jan 21 2021 02:00:00 GMT+0200 (Eastern European Standard Time)
```

toy replied:
moment.js has [strict mode](https://momentjs.com/guides/#/parsing/strict-mode/)

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+        }
+      } else {
+        this.startDateTarget.value = this.dateToIso(dates[0]);
+        this.finishDateTarget.value = '';
```
From trying out the feature for a couple of hours, I'd prefer the removal of the finish date not to happen here. This behaviour also differs from the work package date picker. There, the finish date is kept when setting a new start date. I would only remove it if the start date is after the finish date.

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+        this.highlightField(this.finishDateTarget);
+      }
+    } else {
+      this.dateInputFields
```
How does this case happen? I tried to have the execution stop at a breakpoint I set in here but did not manage to run into it by clicking around in the UI.

dombesz replied:
It happens on the first phase that has both start and finish dates editable, admittedly the `if (!field.disabled) {` can be removed.
<img width="792" alt="image" src="https://github.com/user-attachments/assets/96bf6a73-1dc8-470e-a793-eb04850fe338" />

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
-    return element.classList.contains('active');
+  private clearHighLight() {
+    this.dateInputFields
+        .forEach((el) => el.classList.remove('op-datepicker-modal--date-field_current'));
```
Since the 'op-datepicker-modal--date-field_current' class is references in three methods, I think it safer to move it into a class constant.

dombesz replied:
Done. https://github.com/opf/openproject/blob/64712f0ee03341161b74396458801e2bf8049455/frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts#L53

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+    const dates = event.detail.dates;
+
+    if (dates.length === 1) {
+      if ((this.highlightedField === this.finishDateTarget) || this.startDateTarget.disabled) {
```
When is the second part of the condition (`this.startDateTarget.disabled`) necessary? From clicking around, I couldn't get to a case where the first part is false and only the second part would be true.

dombesz replied:
It is being activated when clicking the calendar on a freshly opened dialog of a successor phase with a finish date already saved.

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+    }
+
+    this.updateFlatpickrCalendar();
+    this.previewForm();
```
This method is complicated to read. But rather than trying to come up with a proposal on how to improve the readability I first have a number of questions regarding it.

dombesz replied:
It got a bit simpler having applied all the modifications.

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+    window.setTimeout(() => {
+      // For mobile, we have to make sure that the active field is scrolled into view after the keyboard is opened
+      field.scrollIntoView(true);
+    }, 300);
```
This leads to an irritating behaviour on non mobil resolutions where the modal is scrolled unnecessarily:

![Kapture 2025-05-21 at 09 31 32](https://github.com/user-attachments/assets/d177ca66-9830-4f18-99ca-eabe8611d60e)

which also sometimes happens without the user clicking into the field:

![Kapture 2025-05-21 at 09 43 03](https://github.com/user-attachments/assets/ba98b1cc-b632-4b7f-8b91-bb199d5070d4)

Maybe this can be limited to only happen on mobile. Or, we figure out how to prevent scrolling in this element by setting the height of the modal to be the same as the height of its content. Because even when not clicking in the input, the behaviour can be triggered which is irritating:

![Kapture 2025-05-21 at 09 37 17](https://github.com/user-attachments/assets/42d3cd3b-8a38-4de3-b593-abefa95aaa4c)

dombesz replied:
Fixed, the main issue was the presence of the scrollbar.

ulferts commented on frontend/src/stimulus/controllers/dynamic/overview/project-life-cycles-form.controller.ts:
```diff
@@ -28,19 +28,170 @@
+    if (isActiveElement && ['value', 'class'].includes(attributeName)) {
+      event.preventDefault();
+    }
+  }
```
Can we extend this to also prevent the morphing of the 'Save' button which flickers with every click on a date:

![Kapture 2025-05-21 at 09 38 28](https://github.com/user-attachments/assets/1fce5b15-c400-4772-9bc6-befb09d26e23)

dombesz replied:
Disabling the save button was added intentionally by providing the submitter button for the form submission. Removing it, since it is confusing.
https://github.com/opf/openproject/blob/32dc0990418725b5f369ef491dcbd9bd523d6c10/frontend/src/stimulus/controllers/form-preview.controller.ts#L46-L47

ulferts commented on app/forms/projects/life_cycles/form.rb:
```diff
@@ -37,42 +43,110 @@ def qa_field_name
+    def value(field_name)
+      case field_name
+      when :start_date
+        model.default_start_date || model.start_date_before_type_cast
```
This leads to a sometimes weird behaviour. The scenario is as follows (maybe edge case):
* Have three phases with dates set (A, B, C)
* Remove the finish date of the phase in the middle (B)
* => Both dates are no longer displayed in the overview page for the phase whose date was just removed. The dates of the phase after that (C) remain unchanged.
* When clicking on the dates of the last phase (C), to open the date picker for it
* => the start date is not the one shown on the overview page but rather the finish date of the first phase (A) + 1

![Kapture 2025-05-21 at 09 54 28](https://github.com/user-attachments/assets/1270ac07-4737-48e5-a0af-815f9047bcd5)

I think keeping the date shown in the overview page, which is the one stored in the database, also for the date shown as the start date in the date picker would be the better approach here. 

While we are on the subject of the overview page. What do you think about always showing the value stored in the database also on the overview page. I mean also in the case where there is only one date set (e.g. by removing the finish date). That would change the behaviour somehow of only showing a complete range but I don't see the harm of it. Maybe also a topic for @marcalcobe to chime in.

dombesz replied:
I discussed both issues with Marc:
1. > What do you think about always showing the value stored in the database also on the overview page. I mean also in the case where there is only one date set (e.g. by removing the finish date).

    We agreed to not show the start date alone, because we may even remove the start date at some point in the future.
2. > The start date is not the one shown on the overview page but rather the finish date of the first phase (A) + 1

    This is indeed problematic, since there is no rescheduling happening with a missing previous phase, I came up with the solution of displaying the start date saved in the database. Additionally, since the previous phase does not have a finish date, we cannot align the start date to it, so instead of setting it "automatically" and disabling it, I decided to enable the start field, unless the previous phase has a complete range.

    <details><summary>See the Demo here</summary>
    <p>

    ![May-27-2025 11-39-17](https://github.com/user-attachments/assets/06ff3259-fc44-45a7-8ea6-e94d4654efa0)

    </p>
    </details>

dombesz replied:
During fixing this issue, I found another similar issue, but that seems to be caused by not rescheduling following phases while the current phase without dates is deactivated.

https://community.openproject.org/wp/64292

ulferts commented on frontend/src/app/shared/components/datepicker/wp-date-picker-modal/wp-date-picker-instance.component.ts:
```diff
@@ -220,31 +221,38 @@ export class OpWpDatePickerInstanceComponent extends UntilDestroyedMixin impleme
+          this.isDayDisabled(dayElem),
+        );
+      },
+      minDate: this.minDate,
```
Do you know if it is possible to have a more explicit visualization for the minDate?. What we currently have left me wondering a couple of times as to why I couldn't click on a certain date.

dombesz replied:
We have a predefined style for disabled days, in order to change it I need to consult the UI team first:

https://github.com/opf/openproject/blob/5dde7f8e02b34ee0ae4fb73ec7f15a02d0e2a6e9/frontend/src/global_styles/content/_datepicker.sass#L31-L35

dombesz replied:
I applied the `--fgColor-disabled` to the disabled days, and it makes them more easy to spot.
<img width="653" alt="image" src="https://github.com/user-attachments/assets/b15cf874-fa80-4fee-9fee-794567b1100f" />

Dark mode
<img width="655" alt="image" src="https://github.com/user-attachments/assets/a2c47e21-bdd2-46ab-87e7-b6267e2bc65f" />

dombesz replied:
For now the decision was made to not make any changes, because the contrast points between the enabled and disabled dates are not sufficient. The UX team will further explore possible colours.

ulferts reviewed (changes requested):
Hi @dombesz, solid PR. As is the nature with something like this (and as the work package date picker has shown us), there are quite a number of smaller things we should at least consider to amend. 

I did check iOS as I have a simulator for that running but I have no such thing for android so I did not test that.

On quite a number of the reported issues I don't claim to know the truth so feel free to discuss.

Apart from what I have noted down right at the code there are the following issues:

### Date validation

When inserting invalid dates (or just a string), no error is displayed on saving but rather, the date picker is closed and the value is not applied:

![Kapture 2025-05-21 at 10 15 02](https://github.com/user-attachments/assets/82dbe863-1805-4b67-ac37-c994a9585448)

### Non working days

You already mentioned that this is an issue so I am just adding it to have a complete checklist when retesting this: 

Subsequent phases have their start date set to the next day when it should be the next non working day.

dombesz committed:
Use default attribute setter in the ProjectLifeCycleSteps::SetAttributesService

dombesz committed:
Create ProjectLifeCyclesFormController constant.

dombesz committed:
Do not disable submit button on form preview.

dombesz committed:
Remove field.disabled condition when setting both dates on the datepicker.

dombesz committed:
Use multiple callbacks in work package factories instead of a lambda.

dombesz committed:
Set the duration by setting the attribute not as a callback in order to avoid having a dirty duration attribute.

dombesz committed:
Rearrange ProjectLifeCycleStep::UpdateContracts specs.

dombesz committed:
Remove using the date_range setter in life cycle activation specs.

dombesz committed:
Simplify life cycle form handling of date change

dombesz committed:
Remove Phase#date_range= setter and other unused range methods

dombesz committed:
Rename Phase#range_set? to #date_dange_set?

dombesz committed:
Move phase setter methods to the set attribute service

dombesz committed:
Scroll datepicker inputs into view only on mobile.

dombesz committed:
Move and fix specs related to life cycle attributes and update service.

dombesz committed:
Adjust datepicker disabled days and non working disabled days color to make it stand out better.

dombesz committed:
Always display the stored start date on the phase edit form.

dombesz committed:
Add Day.next_working method

dombesz committed:
Disable start date selection only when the previous phase has a finish date.

HDinger committed:
Introduce a `Overlay--size-large-portrai` class to avoid scrollbars in larger Dialogs

dombesz committed:
Update finish date clearing logic spec.

dombesz committed:
Add expectations for enabled and disabled dates

dombesz committed:
Add validation for incomplete and bad formatted dates

dombesz committed:
Revert "Adjust datepicker disabled days and non working disabled days color to make it stand out better."

This reverts commit fbf399100a4e9e358cb229417d4641437e4effa0.

dombesz committed:
Submit a preview only when valid dates are entered

dombesz committed:
Move the date format validation to the contract level

ulferts reviewed (approved):
Very good @dombesz. Looks and feels really nice now.

I found one [bug related to highlighting](https://community.openproject.org/wp/64322) that can be fixed later on.
